day5/day5.c

```diff
@@ -0,0 +1,94 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define YES 1
+#define NO 0
+
+struct rule {
+	int value;
+	struct rule *next;
+};
+void insert (struct rule *list[], int idx, int value)
+{
+	struct rule *node = malloc(sizeof(struct rule));
+	if (node == NULL)
+		return;
+	node->value = value;
+	node->next = list[idx];
+	list[idx] = node;
+}
+void clear_list(struct rule *list[])
+{
+	int idx;
+	for (idx = 0; idx < 100; idx++) {
+		struct rule *p = list[idx];
+		while (p != NULL) {
+			struct rule *tmp = p;
+			p = p->next;
+			free(tmp);
+		}
+		list[idx] = NULL;
+	}
+}
+int main(void)
+{
+	FILE *f;
+	struct rule *list[100], *print;
+	int i, j, pages[25], size, idx, value, correct, current, cp;
+	int middles = 0;
+	int mcorrected = 0;
+	char line[256], *token;
+
+	for (i = 0; i < 100; i++) {
+		list[i] = NULL;
+	}
```

Comment:
Want to learn a magic trick?
Instead of initializing the whole list with a `for` loop, you can make use of the "zero-initialization" of local variables (stack allocated memory) by doing the following:
```c
struct rule *list[100] = { 0 }
```
you can also use `{ NULL }`, but `{ 0 }` is more common. This trick initialize your structure with a bulk operation (usually `memset`) and not really a `for` loop. Does it offer performance improvements? Not that much, probably neglectable, but it is easier to read ;)